core/src/main/java/com/graphhopper/routing/ch/CHAlgoFactoryDecorator.java

```diff
@@ -251,23 +251,51 @@ public RoutingAlgorithmFactory getDecoratedAlgorithmFactory(RoutingAlgorithmFact
+        PrepareContractionHierarchies edgeBasedPCH = null;
+        PrepareContractionHierarchies nodeBasedPCH = null;
         for (PrepareContractionHierarchies p : getPreparations()) {
             boolean weightingMatches = p.getCHProfile().getWeighting().matches(map);
-            if (p.isEdgeBased() == edgeBased && weightingMatches)
-                return p;
-            else if (weightingMatches)
-                weightingMatchesButNotEdgeBased = true;
-
+            if (weightingMatches) {
+                if (p.isEdgeBased()) {
+                    edgeBasedPCH = p;
+                } else {
+                    nodeBasedPCH = p;
+                }
+            }
             entriesStrs.add(p.getCHProfile().getWeighting() + "|" + (p.getCHProfile().isEdgeBased() ? "edge" : "node"));
         }
 
-        String hint = weightingMatchesButNotEdgeBased
-                ? " The '" + Parameters.Routing.EDGE_BASED + "' url parameter is missing or does not fit the weightings. Its value was: '" + edgeBased + "'"
-                : "";
-        throw new IllegalArgumentException("Cannot find CH RoutingAlgorithmFactory for weighting map " + map + " in entries: " + entriesStrs + "." + hint);
+        if (edgeBasedPCH == null && nodeBasedPCH == null) {
+            throw new IllegalArgumentException("Cannot find CH RoutingAlgorithmFactory for weighting map " + map + " in entries: " + entriesStrs + ".");
+        }
+        if (map.has(Parameters.Routing.EDGE_BASED)) {
+            boolean edgeBased = map.getBool(Parameters.Routing.EDGE_BASED, false);
+            if (edgeBased) {
+                if (edgeBasedPCH != null) {
+                    return edgeBasedPCH;
+                } else {
+                    throw new IllegalArgumentException("Found a node-based CH preparation for weighting map " + map + ", but requested edge-based CH. " +
+                            "You either need to configure edge-based CH preparation or set the '" + Parameters.Routing.EDGE_BASED + "' " +
+                            "request parameter to 'false' (was 'true'). all entries: " + entriesStrs);
+                }
+            } else {
+                if (nodeBasedPCH != null) {
+                    return nodeBasedPCH;
+                } else {
+                    throw new IllegalArgumentException("Found an edge-based CH preparation for weighting map " + map + ", but requested node-based CH. " +
+                            "You either need to configure edge-based CH preparation or set the '" + Parameters.Routing.EDGE_BASED + "' " +
+                            "request parameter to 'true' (was 'false'). all entries: " + entriesStrs);
+                }
+            }
+        } else {
+            // no edge_based parameter was set, we determine the CH preparation based on what is there
+            if (edgeBasedPCH != null) {
+                return edgeBasedPCH;
+            } else {
+                return nodeBasedPCH;
+            }
+        }
```

Comment:
Would it make sense to reduce the nesting a bit? E.g.  (or is this then too unreadable?)
```java
if (edgeBased && edgeBasedPCH != null)
    return edgeBasedPCH;

if (!edgeBased && nodeBasedPCH != null)
    return nodeBasedPCH;

String str = edgeBased? 
  "Found a node-based CH preparation for weighting map " + map + ", but requested edge-based CH :
  "Found an edge-based CH preparation for weighting map " + map + ", but requested node-based CH : 
throw new IllegalArgumentException(str + ". You either need to configure edge-based CH preparation or set the '" + Parameters.Routing.EDGE_BASED 
+ "' parameter to '"+!edgeBased+"' (was '"+edgeBased+"'). all entries: " + entriesStrs);
```

and 

```java
return edgeBasedPCH != null ? edgeBasedPCH : nodeBasedPCH;
```

Reply:
ok yes I did this. there was an error in the error message as well ('need to configure node-based' instead of 'need to configure edge-based' in the second case)